### PR TITLE
Additions for `iproute2`

### PR DESCRIPTION
**Pull request by no92**

### CODE REVIEW

File: options/posix/generic/posix_string.cpp [hunk 1]
```diff
@@ -132,9 +132,8 @@ char *strcasestr(const char *s, const char *pattern) {
 	return nullptr;
 }
 
-char *strdupa(const char *) {
-	__ensure(!"Not implemented");
-	__builtin_unreachable();
+char *strdupa(const char *str) {
+	return strcpy(reinterpret_cast<char *>(alloca(strlen(str) + 1)), str);
```
Comment (ikbenlike) [inline options/posix/generic/posix_string.cpp:136]: is this code actually safe? from `man alloca`:
> This temporary space is automatically freed when the function that called alloca() returns to its caller.

In this case I would expect the destination buffer to be freed upon `strdupa` returning - for the sake of completeness you should write a test case specificially engineered to use the stack anyway, such that improper use of `alloca` will be detected by seeing that the destination buffer was corrupted

**Review by ikbenlike — CHANGES_REQUESTED** [review]

File: options/posix/generic/posix_string.cpp (hunk 1 above)
Comment (ArsenArsen) [reply options/posix/generic/posix_string.cpp:136]: no

**Comment by ikbenlike** [comment]
> Why should we take over the `_GNU_SOURCE` feature test macro? This seems like a glibc specific thing.

would it not be possible for some software to depend on it working as it does? although providing this stuff unconditionally could work as well, I suppose

File: tests/posix/strdupa.c [hunk 2]
```diff
@@ -0,0 +1,14 @@
+#ifndef _GNU_SOURCE
+#define _GNU_SOURCE
+#endif
+#include <assert.h>
+#include <string.h>
+
+int main() {
+	char test[19] = "Hello mlibc World!";
```
Comment (ikbenlike) [inline tests/posix/strdupa.c:8]: would be good to add an expected fail test case that tries to `free` the `strdupa`d buffer

**Review by ikbenlike — CHANGES_REQUESTED** [review]

File: options/internal/generic/ubsan.cpp [hunk 3]
```diff
@@ -238,3 +238,17 @@ void __ubsan_handle_missing_return(UnreachableData *data) {
...
+	SourceLocation loc;
+	SourceLocation attr_loc;
+	int arg_index;
+};
+
+extern "C" [[gnu::visibility("hidden")]]
+void __ubsan_handle_nonnull_arg(NonNullArgData *data) {
+	mlibc::panicLogger()
+		<< LOG_NAME_LOC("null pointer passed to non-null argument", data->loc)
+		<< "argument " <<data->arg_index << " is required to be non-null in "
```
Comment (Dennisbonke) [inline options/internal/generic/ubsan.cpp:252]: You're missing a space between `<<` and `data->arg_index`.

File: options/linux/include/bits/linux/linux_sched.h [hunk 4]
```diff
@@ -4,6 +4,9 @@
 
 // Glibc extension
 int sched_getcpu(void);
+
+#if defined(_GNU_SOURCE)
 int setns(int fd, int nstype);
+#endif // _GNU_SOURCE
```
Comment (Dennisbonke) [inline options/linux/include/bits/linux/linux_sched.h:10]: We should use C style comments here.

File: options/posix/include/bits/posix/posix_string.h [hunk 5]
```diff
@@ -21,10 +22,23 @@ void *memccpy(void *__restrict dest, const void *__restrict src, int c, size_t n
...
+})
+#define strndupa(x, y) ({ \
+	const char *str = (x); \
+	size_t len = strnlen(str, (y)) + 1; \
+	char *buf = alloca(len); \
+	buf[len - 1] = '\0'; \
+	(char *) memcpy(buf, str, len - 1); \
+})
 void *memrchr(const void *, int, size_t);
+#endif // defined(_GNU_SOURCE)
```
Comment (Dennisbonke) [inline options/posix/include/bits/posix/posix_string.h:41]: We should use C style comments here.

File: tests/posix/strdupa.c (hunk 2 above)
Comment (no92) [reply tests/posix/strdupa.c:8]: That would be UB, though.

**Review by Dennisbonke — CHANGES_REQUESTED** [review]
LGTM outside a few styling issues.

File: tests/posix/strdupa.c (hunk 2 above)
Comment (ikbenlike) [reply tests/posix/strdupa.c:8]: yes, hence it's expected to fail

**Review by Dennisbonke — APPROVED** [review]
LGTM

**Pull request merged**